src/Microsoft.DotNet.Build.CloudTestTasks/UploadClient.cs

```diff
@@ -0,0 +1,196 @@
+// </summary>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    #region
+
+    using System;
+    using System.Collections;
+    using System.Globalization;
+    using System.IO;
+    using System.Net;
+    using System.Text;
+    using Microsoft.Build.Utilities;
+
+    #endregion
+
+    /// <summary>
+    /// The upload client.
+    /// </summary>
+    public class UploadClient
+    {
+        private TaskLoggingHelper log;
+
+        public UploadClient(TaskLoggingHelper log)
+        {
+            this.log = log;
+        }
+
+        /// <summary>
+        /// The encode block ids.
+        /// </summary>
+        /// <param name="numberOfBlocks">
+        /// The number of blocks.
+        /// </param>
+        /// <param name="lengthOfId">
+        /// The length of id.
+        /// </param>
+        /// <returns>
+        /// The <see cref="string"/>.
+        /// </returns>
+        public string EncodeBlockIds(int numberOfBlocks, int lengthOfId)
+        {
+            var numberOfBlocksString = numberOfBlocks.ToString("D" + lengthOfId);
+
+            var bytes = Encoding.UTF8.GetBytes(numberOfBlocksString);
+            return Convert.ToBase64String(bytes);
+        }
+
+        /// <summary>
+        /// The upload block blob.
+        /// </summary>
+        /// <param name="AccountName">
+        /// The account name.
+        /// </param>
+        /// <param name="AccountKey">
+        /// The account key.
+        /// </param>
+        /// <param name="ContainerName">
+        /// The container name.
+        /// </param>
+        /// <param name="filePath">
+        /// The file path.
+        /// </param>
+        /// <param name="destinationBlob">
+        /// The destination blob.
+        /// </param>
+        public void UploadBlockBlob(
+            string AccountName, 
+            string AccountKey, 
+            string ContainerName, 
+            string filePath, 
+            string destinationBlob)
+        {
+            var resourceUrl = string.Format("https://{0}.blob.core.windows.net/{1}", AccountName, ContainerName);
+
+            var filename = destinationBlob;
+            filename = filename.Replace("\\", "/");
+            var blobUploadUrl = resourceUrl + "/" + filename;
+            var size = (int)new FileInfo(filePath).Length;
+            var blockSize = 4 * 1024 * 1024;
+            var bytesLeft = size;
+            var blockIds = new ArrayList();
+            var numberOfBlocks = (size / blockSize) + 1;
+            var countForId = 0;
+            var f = new FileStream(filePath, FileMode.Open);
+            var offset = 0;
+
+            while (bytesLeft > 0)
+            {
+                var nextBytes = (bytesLeft < blockSize) ? bytesLeft : blockSize;
+                var fileBytes = new byte[blockSize];
+                var read = f.Read(fileBytes, 0, nextBytes);
```

Comment:
Technically, "Read" doesn't _have_ to return the number of bytes you requested. Though I don't know of any implementation of FileStream that doesn't.  You should probably validate that "read" and "nextBytes" are the same after this line with some sort of Exception or Error log or something, so if it weirdly fails in the future, we'll see why.